### PR TITLE
Fix Windows dev workflow and worker runner

### PR DESCRIPTION
**Pull request by Kraftis**
## Summary
- Makes `pnpm run migrate` and `pnpm run dev` cross-platform on Windows and POSIX.
- Adds a long-running worker runner with cron scheduling aligned to Kubernetes Helm defaults.
- Updates Docker Compose to run `runner.js` per worker service instead of restart loops.
- Keeps one-shot worker commands for tests, CI, and CronJobs.
- Moves pnpm overrides to `pnpm-workspace.yaml`.

### Review follow-ups:
- Add root `.env.example` for local onboarding (optional; defaults apply without it).
- Improve Windows graceful shutdown in `dev.js` / `run-with-env.js` (`taskkill /t`, bounded hard stop).
- Fix worker CLI entrypoint detection on Windows via `is-node-entrypoint.js`.
- Fix auto-sync / endpoint-check entrypoint exit race (async IIFE + try/catch).
- Validate impossible cron expressions early (`validateCronFeasibility`).
- Cap cron lookahead to 35 days and add 30s shutdown timeout for in-flight runs.
- Set `TZ=UTC` on all worker Compose services and in `Dockerfile.worker`.
- Wire `WORKER_*_INTERVAL_MS` through Compose with empty defaults (cron stays default; set `*_CRON=interval` + value to opt in).
- Document cron subset, Docker CMD vs Compose layout, and `WORKER_EXIT_ON_ERROR` observability.
- 
### Local dev ergonomics:
- `pnpm dev` starts PostgreSQL in Docker and runs DB migrations when the API starts (no separate migrate step).
- `pnpm dev:noDB` skips Docker Postgres; run `pnpm run migrate` first when using an external database.
- Apply local dev defaults from `.env.example` when no root `.env` exists (Postgres credentials stay in sync).
- `pnpm dev:ports:check` detects port conflicts with `tokentimer-enterprise` `demo:local` / `dev-mock-api` on `:4000` and `:5173`.
- Add `deploy/compose/docker-compose.postgres.yml` for Postgres-only local dev.

## Test plan

- [X] `pnpm test:unit` (39 tests pass locally)
- [X] `pnpm --filter @tokentimer/worker build`
- [X] `docker compose -f deploy/compose/docker-compose.yml config`
- [X] Windows: `pnpm dev`, Ctrl+C stops child process tree without orphan Node processes
- [X] Windows: one-shot workers (`queue-manager`, `delivery-worker`, etc.) start correctly
- [X] Compose: workers schedule on cron defaults (discovery */5, delivery 1/5, weekly Mon 9:00 UTC)
- [X] Optional interval mode: set `WORKER_*_CRON=interval` + matching `*_INTERVAL_MS` in `.env`
- [X] Runner rejects impossible crons (e.g. `0 9 31 2 *`) at startup

### CODE REVIEW

File: apps/worker/package.json [hunk 1]
```diff
@@ -6,9 +6,20 @@
   "type": "module",
   "main": "src/queue-manager.js",
   "scripts": {
+    "start": "node src/queue-manager.js",
```
Comment (tokentimerch) [inline apps/worker/package.json]: ### High priority: worker `start` is inconsistent with the new runner

Compose now runs `src/runner.js <worker>`, but this package still defines:

```json
"start": "node src/queue-manager.js"
```

while the runner is only used by `"dev"` / `dev:*`. Root `start:worker` delegates here, so it runs one-shot discovery only.

**Impact:** deployments not using updated Compose commands may start only discovery once and exit, or miss delivery/auto-sync/endpoint/digest workers.

**Suggested fix:** make the long-running entrypoint the default, e.g.

```json
"start": "node src/runner.js all"
```

Keep one-shot scripts under `start:discovery`, `start:delivery`, etc. Also update `deploy/compose/Dockerfile.worker` `CMD` or document that the image default is intentionally one-shot.

File: package.json [hunk 2]
```diff
@@ -4,18 +4,18 @@
...
+    "dev:api": "node scripts/run-with-env.js pnpm --filter @tokentimer/api dev",
+    "dev:worker": "node scripts/run-with-env.js pnpm --filter @tokentimer/worker dev",
+    "dev:dashboard": "node scripts/run-with-env.js pnpm --filter @tokentimer/dashboard dev",
     "build": "pnpm run build:api && pnpm run build:worker && pnpm run build:dashboard",
     "build:api": "pnpm --filter @tokentimer/api build",
     "build:worker": "pnpm --filter @tokentimer/worker build",
     "build:dashboard": "pnpm --filter @tokentimer/dashboard build",
     "start": "pnpm run start:api",
     "start:api": "pnpm --filter @tokentimer/api start",
     "start:worker": "pnpm --filter @tokentimer/worker start",
```
Comment (tokentimerch) [inline package.json:20]: Related to worker entrypoint inconsistency: `start:worker` still maps to `@tokentimer/worker start`, which resolves to one-shot `queue-manager.js`, not `runner.js all`. Consider pointing this at the runner or renaming to make one-shot vs long-running explicit.

File: scripts/process-utils.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      {
+        stdio: "ignore",
+        windowsHide: true,
+      },
+    );
+    killer.on("error", () => {});
+    return;
+  }
+
+  child.kill("SIGTERM");
```
Comment (tokentimerch) [inline scripts/process-utils.js]: ### High priority: `killProcessTree()` is not a tree kill on Unix

Windows uses `taskkill /t /f` (correct). On POSIX this only calls `child.kill("SIGTERM")` on the direct child.

Both `scripts/dev.js` and `scripts/run-with-env.js` spawn `pnpm` and rely on this for Ctrl-C cleanup.

**Impact:** on Linux/macOS, Ctrl-C may kill `pnpm` but leave grandchildren (Vite, API dev server, worker runner) orphaned.

**Suggested fix:** spawn children in a detached process group on POSIX and kill `-pid`, or use a proven tree-kill helper. Add a test that spawns child → grandchild and verifies both exit after shutdown.

File: scripts/load-root-env.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+const fs = require("fs");
+const path = require("path");
+
+function unquote(value) {
```
Comment (tokentimerch) [inline scripts/load-root-env.js:54]: ### High priority: custom `.env` parser mishandles quoted values with inline comments

`unquote()` strips inline comments only when the value is not detected as quoted. For:

```env
API_URL="http://localhost:4000" # local API
```

the raw value does not end with `"`, so the parser strips the comment and returns `"http://localhost:4000"` **including the quotes**. That can leak into `API_URL` / `VITE_API_URL`.

**Suggested fix:** use `dotenv` directly, or strip comments before quote detection. Add unit tests for:

```env
A="value" # comment
B='value' # comment
C=value # comment
D="value # not comment"
```

File: apps/worker/src/runner.js [hunk 5]
```diff
@@ -0,0 +1,373 @@
+#!/usr/bin/env node
+import { pathToFileURL } from "url";
+import { pool } from "./db.js";
+import { logger } from "./logger.js";
+import { queueDiscoveryJob } from "./queue-manager.js";
```
Comment (tokentimerch) [inline apps/worker/src/runner.js]: ### Medium priority: eager imports of every worker module

All worker modules are imported at top level even when running a single worker (`node src/runner.js delivery`).

**Impact:** an unrelated worker import-time failure breaks single-worker mode; enterprise/core override assumptions may be more brittle.

**Suggested fix:** lazy-load only the selected worker (dynamic `import()` factories keyed by worker name).

File: scripts/dev.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+let shuttingDown = false;
+
+const services = [
+  {
+    name: "api",
+    args: ["--filter", "@tokentimer/api", "dev"],
+  },
+  {
+    name: "worker",
+    args: ["--filter", "@tokentimer/worker", "dev"],
```
Comment (tokentimerch) [inline scripts/dev.js:24]: ### Medium priority: `pnpm dev` now starts all workers on startup

`dev.js` starts the worker via `@tokentimer/worker dev` → `runner.js all`. Defaults run auto-sync and endpoint-check on startup.

**Impact:** a developer pointing at a non-local DB/API may immediately trigger auto-sync scans or endpoint checks. Meaningful behavior change vs just starting dev servers.

**Suggested fix:** document clearly in `DEVELOPMENT.md` / `QUICKSTART.md`, or use safer dev defaults e.g. `WORKER_AUTO_SYNC_RUN_ON_START=false` and `WORKER_ENDPOINT_CHECK_RUN_ON_START=false` in dev-oriented env examples.

File: scripts/dev.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    args: ["--filter", "@tokentimer/dashboard", "dev"],
+  },
+];
+
+function stopAll(exitCode) {
+  if (shuttingDown) return;
+  shuttingDown = true;
+
+  for (const child of children.values()) {
+    killProcessTree(child);
```
Comment (tokentimerch) [inline scripts/dev.js:37]: On Windows, `stopAll()` uses `killProcessTree()` → `taskkill /f`, which can terminate the worker runner before `pool.end()` runs. Commit `ac8bc45` fixed SIGINT propagation for `run-with-env.js`, but full-stack `pnpm dev` still hard-kills the tree. Consider graceful shutdown propagation for the worker runner.

File: tests/unit/worker-runner.test.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
+"use strict";
```
Comment (tokentimerch) [inline tests/unit/worker-runner.test.js:1]: ### Medium priority: test coverage is mostly helper-level

Useful coverage for interval parsing, overlap prevention, and weekly-digest run-on-start. Missing failure-prone paths:

- actual process spawning / Ctrl-C cleanup
- `.env` parsing edge cases
- `runner.js stop()` / `--once all`
- Dockerfile / package / Compose entrypoint consistency

**Suggested before merge:**

```bash
pnpm install --frozen-lockfile
pnpm test:unit
pnpm --filter @tokentimer/worker build
docker compose -f deploy/compose/docker-compose.yml config
```

Also: tests import full `runner.js` → `db.js` → `pg`, coupling helpers to the full module graph.

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -2,114 +2,166 @@
...
-**Benefits:**
+Stop the full local stack with `Ctrl+C`.
 
-- Instant hot reload for frontend changes
-- No Docker rebuild needed
-- Backend still in Docker (easier database setup)
+## Environment
 
-### All Local
+Create `.env` in the repository root:
```
Comment (tokentimerch) [inline DEVELOPMENT.md]: Docs instruct creating `.env` at repo root, but there is no root `.env.example` (full reference is only in `deploy/compose/.env.example`). Consider adding a root `.env.example` or linking to the compose example to reduce onboarding friction.

File: apps/worker/queue-architecture.md [hunk 10]
```diff
@@ -70,15 +70,16 @@ CREATE TABLE alert_delivery_log (
...
-Schedules are configurable via Helm values (`worker.cronjobs.*`) or Docker Compose environment.
-
-| Job | Default schedule | Helm value |
-|-----|-----------------|------------|
-| Alert Discovery | `*/5 * * * *` | `worker.cronjobs.discovery.schedule` |
-| Alert Delivery | `1/5 * * * *` | `worker.cronjobs.delivery.schedule` |
-| Weekly Digest | `0 9 * * 1` | `worker.cronjobs.weeklyDigest.schedule` |
-| Endpoint Check | `*/1 * * * *` | `worker.cronjobs.endpointCheck.schedule` |
-| Auto Sync | `0 * * * *` | `worker.cronjobs.autoSync.schedule` |
+Kubernetes uses CronJobs by default. Docker Compose uses the long-running
```
Comment (tokentimerch) [inline apps/worker/queue-architecture.md]: Compose runner intervals differ from Helm cron defaults (e.g. discovery 60s vs */5, weekly digest 24h poll vs Mon 9am). Documented here, but worth confirming production Compose users understand weekly digest is "poll every 24h + DB dedup", not Monday morning cron semantics.

File: deploy/compose/docker-compose.yml [hunk 11]
```diff
@@ -168,6 +168,9 @@ services:
       ENABLE_METRICS: ${ENABLE_METRICS:-false}
       PUSHGATEWAY_URL: ${PUSHGATEWAY_URL:-}
       ENVIRONMENT_SUFFIX: ${ENVIRONMENT_SUFFIX:-}
+      WORKER_DISCOVERY_INTERVAL_MS: ${WORKER_DISCOVERY_INTERVAL_MS:-60000}
+      WORKER_RUN_ON_START: ${WORKER_RUN_ON_START:-true}
+      WORKER_EXIT_ON_ERROR: ${WORKER_EXIT_ON_ERROR:-false}
```
Comment (tokentimerch) [inline deploy/compose/docker-compose.yml:175]: ### Lower priority: silent failure mode

`WORKER_EXIT_ON_ERROR=false` means recurring job failures are logged but the container keeps running. That avoids restart-looping on job errors, but production needs external signals (metrics, last-success gauges, alerts) — otherwise a broken worker can look healthy to Docker.

File: deploy/compose/docker-compose.yml [hunk 12]
```diff
@@ -298,6 +304,9 @@ services:
       ENABLE_METRICS: ${ENABLE_METRICS:-false}
       PUSHGATEWAY_URL: ${PUSHGATEWAY_URL:-}
       ENVIRONMENT_SUFFIX: ${ENVIRONMENT_SUFFIX:-}
+      WORKER_WEEKLY_DIGEST_INTERVAL_MS: ${WORKER_WEEKLY_DIGEST_INTERVAL_MS:-86400000}
```
Comment (tokentimerch) [inline deploy/compose/docker-compose.yml]: Weekly digest default interval is `86400000` ms (daily) while the worker is named weekly-digest. Likely fine because `weeklyDigestJob()` deduplicates via `weekly_digest_log` per `week_start_date`, but please confirm a daily runner cannot send duplicate weekly digests.

File: pnpm-workspace.yaml [hunk 13]
```diff
@@ -1,3 +1,27 @@
 packages:
   - "apps/*"
   - "packages/*"
+
+overrides:
```
Comment (tokentimerch) [inline pnpm-workspace.yaml:5]: Moving `overrides` and `onlyBuiltDependencies` here looks correct for pnpm 10. Please confirm `pnpm install --frozen-lockfile` passes in CI with the lockfile unchanged/regenerated as needed.

**Review by tokentimerch — CHANGES_REQUESTED** [review]
Thanks for the Windows dev and worker-runner work: the direction is good. I reviewed the PR head changes and would not merge as-is without resolving a few inconsistencies.

**Summary:** The PR improves Windows dev ergonomics and moves Compose away from restart-as-scheduler toward a proper long-running runner. The runner design (reuse job functions, overlap guard, `--once`, pool lifecycle for discovery/delivery) is solid. Before merge, please align worker entrypoints, fix POSIX process-tree cleanup, and add tests for `.env` parsing and shutdown behavior.

See inline comments for details. Additional notes not tied to a changed line:

- **`deploy/compose/Dockerfile.worker`** (unchanged in this PR): default `CMD` is still `node apps/worker/src/queue-manager.js`. Compose overrides it, but direct image users still get one-shot discovery.
- **Compose vs Helm schedules:** Compose defaults (60s discovery, 30s delivery, 5m auto-sync, 24h weekly digest) differ from Helm cron semantics. Weekly digest dedup in `weekly_digest_log` prevents spam, but production Compose users may expect Monday 9am, not a rolling 24h poll.
- **Windows `import.meta.url` checks** in one-shot worker entrypoints (`queue-manager.js`, etc.) still use `new URL(process.argv[1], "file://")` while `runner.js` correctly uses `pathToFileURL`.
- **Pre-existing `.catch().then()` race** in `auto-sync-worker.js` / `endpoint-check-worker.js` entrypoints (unaffected by runner mode, but still risky for CronJobs).

**Verdict:** Good direction; please address high-priority entrypoint and cleanup issues first.

**Review by tokentimerch — APPROVED** [review]
